src/child.cpp

```diff
@@ -0,0 +1,473 @@
+#include <unistd.h>
+
+int clients;
+Mutex *cmutex;
+
+child_t childs[CLIENTS_MAX + 1];
+
+int child_find_unused() {
+  //  printf("Child find unused...");
+  int idx = -1;
+  int i = 0;
+  cmutex->lock();
+  while (idx < 0 && i < (CLIENTS_MAX + 1)) {
+    if (!childs[i].arg.used) {
+      childs[i].arg.used = 1;
+      idx = i;
+    }
+    i++;
+  }
+  cmutex->unlock();
+  return idx;
+}
+
+void child_list_clear() {
+  for (int i = 0; i < (CLIENTS_MAX + 1); i++)
+    childs[i].arg.used = 0;
+}
+
+const char helpstr[] = "QSCAND: valid commands:\n\
+close           close connection\n\
+list|scanbus    list available devices\n\
+dinfo           show device info (have to set device first)\n\
+minfo           show media info (have to set device first)\n\
+run             run test (have to set device and test type first)\n\
+\n\
+get             get current parameters\n\
+set <PAR>=<VAL> set parameter. PAR can be:\n\
+                dev   - set device, VAL should be like '/dev/hdX' or '/dev/sdX'\n\
+                test  - set test type, VAL should be one of folowing:\n\
+                        rt   - read transfer rate\n\
+                        wt   - write transfer rate\n\
+                        errc - error correction test\n\
+                        jb   - jitter/asymmetry test\n\
+                        ft   - focus/tracking test\n\
+                        ta   - time analyser\n\
+                speed - set test speed, VAL should be integer\n\
+                simul - set test speed, VAL should be 0 or 1 (default: 1)\n\
+";
+
+const int helpstr_sz = sizeof(helpstr);
+
+enum qscan_mode { none = 0, scanbus = 1, scan, dinfo, minfo, help };
+
+char tchar = -1;
+
+enum fdtype_t { FD_PIPE = 1, FD_SOCKET = 2 };
+
+// int readline(int fd, char *buf, int maxlen, fdtype_t fdtype = FD_PIPE)
+int readline(int fd, char *buf, int maxlen, fdtype_t fdtype) {
+  int cnt = 0;
+  char *cbuf = buf;
+  int r;
+  int sret;
+  fd_set rd_set;
+  timeval tv;
+
+  FD_ZERO(&rd_set);
+
+  //  printf("tchar=%02X\n", tchar);
+  if (tchar >= 0) {
+    cbuf[0] = tchar;
+    cnt++;
+    cbuf++;
+    tchar = -1;
+  }
+  while (!term && (cnt < (maxlen - 1))) {
+    FD_SET(fd, &rd_set);
+    tv.tv_sec = 1;
+    tv.tv_usec = 0;
+#ifdef _WIN32
+    if (fdtype == FD_SOCKET) {
+#endif
+      sret = select(fd + 1, &rd_set, NULL, NULL, &tv);
+#ifdef _WIN32
+      errno = GetLastError();
+    } else {
+      sret = 1;
+      errno = 0;
+    }
+#endif
+    //      printf("select(%d): %s\n", fd, strerror(errno));
+    if (sret < 0) {
+#ifndef _WIN32
+      if (debug) {
+        if (debug && !daemonized)
+          std::cout << "readline() " << fd << " select: " << strerror(errno)
+                    << "\n";
+      }
+#endif
+      if (errno == EINTR)
+        continue;
+      return -1;
+    } else if (sret > 0 && FD_ISSET(fd, &rd_set)) {
+#ifdef _WIN32
+      if (fdtype == FD_SOCKET)
+        r = recv(fd, cbuf, 1, 0);
+      else
+#endif
+        r = read(fd, cbuf, 1);
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)